metadata/src/main/java/org/apache/kafka/controller/OffsetControlManager.java

```diff
@@ -212,10 +213,14 @@ long lastStableOffset() {
     }
 
     /**
-     * @return the transaction start offset, or -1 if there is no transaction.
+     * @return the transaction start offset if present, or empty if there is no transaction.
      */
-    long transactionStartOffset() {
-        return transactionStartOffset;
+    OptionalLong transactionStartOffset() {
```

Comment:
I’m not loving this idea of OffsetControlManager.transactionStartOffset returning an OptionalLong. Seems inconsistent with the other accessors.

Maybe let’s just return a regular long? I had some vague efficiency related reasons for preferring the negative / non-negative convention (to be fair, haven’t benchmarked)

Reply:
I think it's a bit of a leaky abstraction to actually return -1 from this class, but I'm fine leaving it as-it (or addressing in another PR)

Reply:
thanks